File: xaiop-sdk/python/src/xaiop/phase_encode.py
"""Encode one skeleton/module phase for WebSocket push."""

from __future__ import annotations

from typing import Any

from .encode import encode_sync


def encode_phase_json(
    key: str,
    value: Any,
    *,
    final: bool = False,
    encode_options: dict[str, Any] | None = None,
) -> str:
    if not isinstance(key, str) or not key:
        raise TypeError("phase key must be a non-empty string")
    opts = {**(encode_options or {}), "dot_policy": "none"}
    wire = encode_sync({key: value}, **opts)
    if not final:
        wire = _ensure_trailing_newline(wire) + ".\n"
    return wire


def encode_phase_object(
    obj: dict[str, Any],
    *,
    final: bool = False,
    encode_options: dict[str, Any] | None = None,
) -> str:
    if obj is None or not isinstance(obj, dict):
        raise TypeError("phase object must be a plain object")
    opts = {**(encode_options or {}), "dot_policy": "none"}
    wire = encode_sync(obj, **opts)
    if not final:
        wire = _ensure_trailing_newline(wire) + ".\n"
    return wire


def _ensure_trailing_newline(wire: str) -> str:
    return wire if wire.endswith("\n") else f"{wire}\n"

File: xaiop-sdk/go/xaiop/phase_encode.go
package xaiop

import "fmt"

// PhaseEncodeOptions controls trailing phase '.' for skeleton push.
type PhaseEncodeOptions struct {
	Final  bool // when true, omit trailing ".\n"
	Encode EncodeOptions
}

// PhaseEncodeJSON encodes a single Diff/phase payload as XAIOP with
// DotPolicy forced to "none" (Node phaseEncode).
func PhaseEncodeJSON(value any, opts EncodeOptions) (string, error) {
	opts.DotPolicy = "none"
	if opts.Style == "" {
		opts.Style = "relative"
	}
	if !opts.TrailingNewline {
		opts.TrailingNewline = true
	}
	return Encode(value, opts)
}

// PhaseEncodeObject is an alias for PhaseEncodeJSON for object roots.
func PhaseEncodeObject(value map[string]any, opts EncodeOptions) (string, error) {
	return PhaseEncodeJSON(value, opts)
}

// PhaseEncodeKeyValue encodes {key: value} as one skeleton phase (append ".\n" unless Final).
func PhaseEncodeKeyValue(key string, value any, opts PhaseEncodeOptions) (string, error) {
	if key == "" {
		return "", fmt.Errorf("phase key must be a non-empty string")
	}
	return finishPhaseEncode(map[string]any{key: value}, opts)
}

// PhaseEncodeObjectPhase encodes a plain object as one skeleton phase.
func PhaseEncodeObjectPhase(obj map[string]any, opts PhaseEncodeOptions) (string, error) {
	if obj == nil {
		return "", fmt.Errorf("phase object must be a plain object")
	}
	return finishPhaseEncode(obj, opts)
}

func finishPhaseEncode(value any, opts PhaseEncodeOptions) (string, error) {
	enc := opts.Encode
	enc.DotPolicy = "none"
	if enc.Style == "" {
		enc.Style = "relative"
	}
	enc.TrailingNewline = true
	wire, err := Encode(value, enc)
	if err != nil {
		return "", err
	}
	if opts.Final {
		return wire, nil
	}
	if !hasTrailingNewline(wire) {
		wire += "\n"
	}
	return wire + ".\n", nil
}

func hasTrailingNewline(s string) bool {
	return len(s) > 0 && s[len(s)-1] == '\n'
}
